index.bs

```diff
@@ -10651,6 +10657,155 @@ The [=remote end steps=] given |session| and |command parameters| are:
 
 </div>
 
+## The autofill Module ## {#module-autofill}
+
+The <dfn export for=modules>autofill</dfn> module contains functionality for
+saving and triggering autofill.
+
+A [=/browsing context=] has a <dfn>autofill store for testing</dfn>, an
+<a>ordered map</a>.
+
+### Definition ### {#module-autofill-definition}
+
+[=remote end definition=]
+
+<pre class="cddl remote-cddl">
+
+InputCommand = (
+  autofill.save //
+  autofill.trigger
+)
+</pre>
+
+### Commands ### {#module-autofill-commands}
+
+#### The autofill.save Command ####  {#command-autofill-save}
+
+The <dfn export for=commands>autofill.save</dfn> command saves a set of field
+name values to the [=/browsing context=]'s [=autofill store for testing=].
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+    <pre class="cddl remote-cddl">
+      autofill.save = (
+        method: "autofill.save",
```

Comment:
I think it would be preferred to save the data by exercising the browser logic based on a filled out form instead of adding autofill fields to a test only storage.

Reply:
For example, if a form is split across multiple iframes, the saving might work or not work in the same way

Reply:
> I'm guessing this would be multiple forms that act as one (through JS calls across the same-origin iframes). Do I understand you correctly?

implementations might also use cross-origin iframes and from autofill perspective that would be one form too.

I guess that if the goal of the proposed save method is not to actually test the saving functionality of a given autofill implementation and just to seed the implementation storage with test data, the current proposal works too.

Reply:
In that case, do we need to provide a list of forms and iframes as part of the API?



> I guess that if the goal of the proposed save method is not to actually test the saving functionality of a given autofill implementation and just to seed the implementation storage with test data, the current proposal works too.

That's indeed my initial goal, but it'd be good if the API enabled testing of the saving functionality as well. (as presumably, this is also something that sites can do correctly or incorrectly)

Reply:
> That's indeed my initial goal, but it'd be good if the API enabled testing of the saving functionality as well. (as presumably, this is also something that sites can do correctly or incorrectly)

I guess this is interesting but perhaps to a lesser extend: it is very likely that the data was stored by filling out forms previously on other websites or entering the data manually into the browser settings.

Reply:
> In that case, do we need to provide a list of forms and iframes as part of the API?

I think we can just provide a SharedReference to the form element that is to be saved to autofill?

Reply:
@OrKoN @yoavweiss @theindra throwing in another potential approach here, to avoid an explicit `save` command: What if we passed in the expected state of the autofill "Storage" on the implementation on every `trigger` call? Something like (pseudo-code, don't focus on naming of the variables :-) ):
`autofill.trigger(onMyForm, withAutofillState)`
where `onMyForm` is the form I want to see the fields being filled, and `withAutofillState` being a key/value hash that describes the current state of autofill for the test.

This more functional approach makes it also very clear that the state of autofill is only expected to be preserved during the context in which the test runs, i.e. no sideeffects from non-cleared autofill storage state compared to an explicit `save()` that kind of implies a `clear()` after the test ran.

Reply:
Yes, this is the approach I [mentioned](https://github.com/w3c/webdriver-bidi/pull/706#discussion_r1613177353) and we went this direction in CDP https://chromedevtools.github.io/devtools-protocol/tot/Autofill/#method-trigger

Reply:
I'd also see value in both having a "shortcut" to pass explicit autofill state into the test (when testing filling forms), as well as creating a form to test the saving functionality of autofill. For me, saving properly, and filling properly, are 2 different things, and if a test fails for my implementation, being able to clearly distinguish whether saving failed, or filling failed, makes sense.

I can imaging the test suite testing filling the fields uses the "pass state into the test", to be independent of the saving functionality. When testing saving, we'd obviously have to use the form approach. This would clearly separate the 2, and would highlight clearly what part failed.

Reply:
@OrKoN I've changed the spec to not have `autofill.save` anymore, and instead pass in the autofill fields for each test. Curious to get your feedback on it.